### PR TITLE
[SPARK-13792][SQL] Limit logging of bad records in CSVRelation

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?

Currently in `PERMISSIVE` and `DROPMALFORMED` modes we log any record that is going to be ignored. This can generate a lot of logs with large datasets. This pr is to log the parts of malformed records and the number of subsequent records for each partition.
This adds two options as follows;

```
sqlContext.read
  .format("csv")
  .option("mode", "COUNTMALFORMED")
  .option("maxStoredMalformedPerPartition", 3)
  .load("test.csv").show
```

A logging message is;

```
16/04/05 16:42:12 WARN CSVRelation: # of total malformed lines: 25
3 malformed lines extracted and listed as follows;
ab ccc ddd ddd
ab ccc ddd ddd
...
```
## How was this patch tested?

Manual tests done

### CODE REVIEW

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
@falaki Could you check this to match your intent?

**Comment by falaki** [comment]
I don't think we need a separate parsing mode for this feature. This feature can be part of all existing modes.

**Comment by maropu** [comment]
okay and fixed. Please re-check again? cc: @falaki

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 1]
```diff
@@ -109,21 +141,45 @@ object CSVRelation extends Logging {
...
+            logWarning("Parse exception. " + s"Dropping malformed line: ${line}")
+            malformedLinesInfo.add(line)
             None
         }
       }
     }
   }
 
-  def parseCsv(
+  def parseCsvInRdd(
```
Comment (falaki) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: This can be private[csv]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 2]
```diff
@@ -109,21 +141,45 @@ object CSVRelation extends Logging {
...
+      val rows = iter.flatMap(parser(_).toSeq)
+      CompletionIterator[InternalRow, Iterator[InternalRow]](rows, {
+        if (malformedLinesInfo.malformedLineNum > 0) {
+          logWarning(s"$malformedLinesInfo")
+        }
+      })
+    }
+  }
+
+  def parseCsvInIterator(
```
Comment (falaki) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 3]
```diff
@@ -33,6 +34,31 @@ import org.apache.spark.sql.catalyst.expressions.GenericMutableRow
 import org.apache.spark.sql.execution.datasources.PartitionedFile
 import org.apache.spark.sql.sources._
 import org.apache.spark.sql.types._
+import org.apache.spark.util.CompletionIterator
+
+/**
+ * Stores and counts malformed lines during CSV parsing.
+ */
+private[csv] class MalformedLinesInfo(maxStoreMalformed: Int) extends Serializable {
```
Comment (falaki) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: I think you can put this in a util file under csv/. Also would be good to unit test it.

Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: okay, I'll add tests.

**Comment by maropu** [comment]
okay, fixed cc:  @falaki

**Comment by maropu** [comment]
@falaki ping

**Comment by maropu** [comment]
@falaki ping

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 4]
```diff
@@ -74,12 +76,17 @@ object CSVRelation extends Logging {
 
     (tokens: Array[String]) => {
       if (params.dropMalformed && schemaFields.length != tokens.length) {
-        logWarning(s"Dropping malformed line: ${tokens.mkString(params.delimiter.toString)}")
+        val line = tokens.mkString(params.delimiter.toString)
+        logWarning(s"Dropping malformed line: $line")
```
Comment (falaki) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Do we still want to log this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala [hunk 5]
```diff
@@ -94,6 +94,8 @@ private[sql] class CSVOptions(@transient private val parameters: Map[String, Str
 
   val maxCharsPerColumn = getInt("maxCharsPerColumn", 1000000)
 
+  val maxStoredMalformedPerPartition = getInt("maxStoredMalformedPerPartition", 1)
```
Comment (falaki) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: Since this affects logging, I suggest a name that implies that. How about `maxLogRecordsPerPartition`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 6]
```diff
@@ -109,21 +116,45 @@ object CSVRelation extends Logging {
           Some(row)
         } catch {
           case NonFatal(e) if params.dropMalformed =>
-            logWarning("Parse exception. " +
-              s"Dropping malformed line: ${tokens.mkString(params.delimiter.toString)}")
+            val line = tokens.mkString(params.delimiter.toString)
+            logWarning("Parse exception. " + s"Dropping malformed line: ${line}")
```
Comment (falaki) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: ditto

**Comment by falaki** [comment]
@maropu sorry for delayed response. I seem to have missed the github notification on this.

**Comment by maropu** [comment]
@falaki okay, I'll fix in a day. thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala (hunk 4 above)
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: This output seems to be noisy to me, so I'll remove this.

**Comment by maropu** [comment]
@falaki okay, fixed.

**Comment by maropu** [comment]
@falaki ping

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
@falaki ping

**Comment by maropu** [comment]
@falaki ping

**Comment by maropu** [comment]
@rxin @falaki ping

**Comment by rxin** [comment]
@maropu thanks for working on this. One thing I don't get is that why we are keeping the malformed lines in memory? Can't we just track a counter and stop logging once the number is greater than a threshold?

**Comment by maropu** [comment]
@rxin yea, the current implementation only holds `malformedLineNum` malformed lines on memory: https://github.com/apache/spark/pull/12173/files#diff-18b09be18156e81f965df293a2781aefR31

Anything I misunderstand?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CsvUtils.scala [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+package org.apache.spark.sql.execution.datasources.csv
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Stores and counts malformed lines during CSV parsing.
+ */
+private[csv] class MalformedLinesInfo(maxStoreMalformed: Int) extends Serializable {
+
+  var malformedLines = new ArrayBuffer[String]
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CsvUtils.scala]: do we need this?

Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CsvUtils.scala]: The some logs of malformed lines seem to be useful for users to understand why this is a malformed one. However, removing this is okay to me.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CsvUtils.scala]: You can just log them as we find them, can't we? No need to store them in memory?

Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CsvUtils.scala]: oh, I see. okay, I'll fix this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala [hunk 8]
```diff
@@ -118,9 +120,20 @@ class CSVFileFormat extends FileFormat with DataSourceRegister {
...
-      val parser = CSVRelation.csvParser(dataSchema, requiredSchema.fieldNames, csvOptions)
-      tokenizedIterator.flatMap(parser(_).toSeq)
+      val unsafeRowIterator = {
+        val tokenizedIterator = new BulkCsvReader(lineIterator, csvOptions, headers)
+        CSVRelation.parseCsvInIterator(tokenizedIterator, dataSchema, requiredSchema.fieldNames,
+          csvOptions)
+      }
+
+      // Appends partition values
+      val fullOutput = requiredSchema.toAttributes ++ partitionSchema.toAttributes
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala]: is this change related to the issue you are fixing?
